_posts/2020-07-30-deploy-create-react-app-aws-pulumi.md

```diff
@@ -0,0 +1,351 @@
+---
+layout: post
+title:  "Deploying a Create React App to AWS with pulumi"
+date:   2020-77-39 00:00:00 +0100
+categories: pulumi aws react
+---
+
+# Getting started with create react app
+
+So you have just created your first app in create react app.
+
+https://reactjs.org/docs/create-a-new-react-app.html
+
+You built it, change some source files. It works! Well, it works on your machine. 
+What's next? How do you actually get it out there, running? 
+You can build it but you want to ship it
+
+The aim of this guide is to provide a step-by-step, incremental improvement journey. 
+What's the point in showing you how to build the castle, if it's not clear why one needs anything more than a shack?
+
+This is not just a cookbook recipie that assumes things will work out. 
+When possible, debuig and diagnostic commands are run to make sure things are in the state that they need to be. 
+
+This will be a big journey but if we can break it up into smaller substeps. 
+Iteratively, we'll go from a shack to a castle. 
+
+# Starting with a 'stock' create react app
+
+So what's the starting point of this journey? 
+If you follow the create react app docs, you'll see something like tis 
+
+```console
+npx create-react-app my-app
+cd my-app
+npm build
+```
+
+This gives you a build/ directory – let's have a look inside
+
+```console
+$ ls build/
+asset-manifest.json					logo512.png						service-worker.js
+favicon.ico						manifest.json						static
+index.html						precache-manifest.a6c522ff242ab9465073ffb9aae702c8.js
+logo192.png						robots.txt
+```
+
+But what do we do with that? How do you get your precious creation into the cloud?
+How to ensure that as you push updates to your app clients get the latest version?
+There are some big questions there, in particular which *cache directives* to set for the files in build/. 
+These settings will be super important for the browser. 
+We'll cover all of that later once we have the nuts and bolts ready. 
+
+Can these topics feel confusing and annoying? Hell yeah. 
+Have you ever seen them covered coprehensively in create react app documentation? Hell no. 
+
+## The plan
+
+The first step is getting an AWS account. That's a little basic for this post and perhaps it's better to leave it for the reader to research. 
+
+What will we need from AWS? It is going to be a million services or just a few? For now just two: the app will need a domain (Route 53) and we'll need a place to put the files into (for that pourpose an S3 bucket will do). 
+
+There are further wrinkles. Which caching settings to use for the resources? How to scale the service in the future and how to monitor it? All these we'll be covered here – we'll be building up from the simplest shack to a more robust confgiruation. 
```

Comment:
`We'll cover all these here – building up from the simplest shack to a more robust configuration. `